controllers/webhook/runcompletion_feed.go

```diff
@@ -0,0 +1,100 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	config "github.com/sky-uk/kfp-operator/apis/config/v1alpha5"
+	"github.com/sky-uk/kfp-operator/apis/pipelines"
+	"io"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type EventData struct {
+	// Header is the http request header
+	Header http.Header `json:"header"`
+	// Body is http request body
+	Body json.RawMessage `json:"body"`
+}
+
+type RunCompletionFeed struct {
+	ctx       context.Context
+	upstreams []UpstreamService
+}
+
+func NewRunCompletionFeed(ctx context.Context, endpoints []config.Endpoint) RunCompletionFeed {
+	upstreams := pipelines.Map(endpoints, func(endpoint config.Endpoint) UpstreamService {
+		return NewHttpWebhook(endpoint)
+	})
+	return RunCompletionFeed{
+		ctx:       ctx,
+		upstreams: upstreams,
+	}
+}
+
+func getRequestBody(request *http.Request) ([]byte, error) {
+	if request.Body == nil {
+		return nil, errors.New("request body is nil")
+	}
+	body, err := io.ReadAll(request.Body)
+	request.Body = io.NopCloser(bytes.NewBuffer(body))
```

Comment:
how come we need to do this? go noob here, but when is the close method called?

Reply:
Good call and not your lack on knowledge of go, more mine :) I have missed the defer close call. One thing I have learnt over the years is to ALWAYS CONSUME YOUR RESPONSE! Will correct now

Reply:
Fixed in [588b9f6](https://github.com/sky-uk/kfp-operator/pull/368/commits/588b9f6642048c2a530697486a392a27918a328f)